### ISSUE TITLE
Why does the softmax actication only work on 2D and 3D tensors

### ISSUE THREAD

**Issue by prinsherbert** [fchollet/keras#5064]
Softmax for >=2D tensors seems trivial in both theano and tensorflow using dimension -1 when maxing and summing, or just using the built-in softmax functions:

- [Theano](http://deeplearning.net/software/theano/library/tensor/nnet/nnet.html#theano.tensor.nnet.nnet.softmax)
- [Tensorflow](https://www.tensorflow.org/api_docs/python/nn/classification#softmax)

Nevertheless, [here](
https://github.com/fchollet/keras/blob/master/keras/activations.py#L15) it is enforced that there are 2 or 3 dimensions when usign softmax. Why?

**Comment by gvtulder** [comment]
This is not an answer to your question, but be aware that Theano `T.softmax` uses dimension 1.

**Comment by prinsherbert** [comment]
@gvtulder You are right, but still, Keras can use the code they provide in their documentation:

    e_x = exp(x - x.max(axis=1, keepdims=True))
    out = e_x / e_x.sum(axis=1, keepdims=True)

Where you can specify the desired axis.

**Comment by israelg99** [comment]
Any info about this?

**Comment by prinsherbert** [comment]
I would be surprised if Keras its softmax is RNN-only. In fact, many convolutional networks end up with a 1D feature vector for each image sample, which is then passed to a softmax for classification (Inception, AlexNet, LeNet, and probably a lot more).

**Comment by wosiu** [comment]
Up. Are there any plans to add softmax for 3D tensors (like after convolution layer)?

**Comment by aendrs** [comment]
Have you solved the use of softmax for feature maps?

**Comment by prinsherbert** [comment]
Probably there is a `Lambda(...)` fix for it on Google, frankly I can't remember.
